rust/phoenix-channel/src/login_url.rs

```diff
@@ -0,0 +1,224 @@
+}
+
+impl Zeroize for LoginUrl {
+    fn zeroize(&mut self) {
+        let placeholder = Url::parse("http://a.com").expect("placeholder URL to be valid");
+        let _ = std::mem::replace(&mut self.url, placeholder);
+    }
+}
+
+impl CloneableSecret for LoginUrl {}
+
+impl LoginUrl {
+    pub fn client<E>(
+        url: impl TryInto<Url, Error = E>,
+        firezone_token: &SecretString,
+        device_id: String,
+        device_name: Option<String>,
+        public_key: [u8; 32],
+    ) -> std::result::Result<Self, LoginUrlError<E>> {
+        let external_id = hex::encode(sha2::Sha256::digest(device_id));
+        let device_name = device_name
+            .or(get_host_name())
+            .unwrap_or_else(|| Uuid::new_v4().to_string());
+
+        let url = get_websocket_path(
+            url.try_into().map_err(LoginUrlError::InvalidUrl)?,
+            firezone_token,
+            "client",
+            Some(public_key),
+            Some(external_id),
+            Some(device_name),
+            None,
+            None,
+        )?;
+
+        Ok(LoginUrl {
+            host: parse_host(&url)?,
+            url,
+        })
+    }
+
+    pub fn gateway<E>(
+        url: impl TryInto<Url, Error = E>,
+        firezone_token: &SecretString,
+        device_id: String,
+        device_name: Option<String>,
+        public_key: [u8; 32],
+    ) -> std::result::Result<Self, LoginUrlError<E>> {
+        let external_id = hex::encode(sha2::Sha256::digest(device_id));
+        let device_name = device_name
+            .or(get_host_name())
+            .unwrap_or_else(|| Uuid::new_v4().to_string());
+
+        let url = get_websocket_path(
+            url.try_into().map_err(LoginUrlError::InvalidUrl)?,
+            firezone_token,
+            "gateway",
+            Some(public_key),
+            Some(external_id),
+            Some(device_name),
+            None,
+            None,
+        )?;
+
+        Ok(LoginUrl {
+            host: parse_host(&url)?,
+            url,
+        })
+    }
+
+    pub fn relay<E>(
+        url: impl TryInto<Url, Error = E>,
+        firezone_token: &SecretString,
+        device_name: Option<String>,
+        ipv4_address: Option<Ipv4Addr>,
+        ipv6_address: Option<Ipv6Addr>,
+    ) -> std::result::Result<Self, LoginUrlError<E>> {
+        let url = get_websocket_path(
+            url.try_into().map_err(LoginUrlError::InvalidUrl)?,
+            firezone_token,
+            "relay",
+            None,
+            None,
+            device_name,
+            ipv4_address,
+            ipv6_address,
+        )?;
+
+        Ok(LoginUrl {
+            host: parse_host(&url)?,
+            url,
+        })
+    }
+
+    // TODO: Only temporarily public until we delete other phoenix-channel impl.
+    pub fn inner(&self) -> &Url {
+        &self.url
+    }
+
+    // TODO: Only temporarily public until we delete other phoenix-channel impl.
+    pub fn host(&self) -> &str {
+        &self.host
+    }
```

Comment:
These will eventually be `pub(crate)` once we delete the other phoenix-channel client impl.